dbms/src/Processors/Sources/SourceWithProgress.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+#include <Processors/ISource.h>
+#include <DataStreams/IBlockInputStream.h>
+#include <Common/Stopwatch.h>
+
+namespace DB
+{
+
+/// Adds progress to ISource.
+/// This class takes care of limits, quotas, callback on progress and updating performance counters for current thread.
+class ISourceWithProgress : public ISource
```

Comment:
I think it is ok for now.
We do have sources without progress (e.g. when we dump to something to disk).
Also I will probably simplify it later, by moving progress logic to executor.
I can't do it now because of compatibility with IBlockInputStream.